### PR TITLE
Alma: Improve support for displaying digital and electronic inventory.

### PR DESCRIPTION
**Pull request by EreMaijala**
This is a proposal to extend the holdings display for electronic and digital inventory when provided by the ILS. Includes support for the Demo driver and Alma. This is motivated by the fact that the Alma driver currently displays both in status results but not in the holdings tab.

A couple of notes:

- While Alma distinquishes between digital and electronic material, I'm not sure we need the distinction here.
- We have a lot of stuff in holdingsils.phtml. It all gets reloaded while paging holdings, right?

TODO (after merging)
- [x] Update changelog
- [x] Update driver spec in wiki

### CODE REVIEW

File: languages/fi.ini [hunk 1]
```diff
@@ -159,7 +159,7 @@ By Recent = "Uutuuden mukaan"
 By Region = "Alueen mukaan"
 By Title = "Nimekkeen mukaan"
 By Topic = "Aiheen mukaan"
-Call Number = "Sijainti"
+Call Number = "Hyllypaikka"
```
Comment (demiankatz) [inline languages/fi.ini]: Is this a generic change that should be applied to existing translations, or is it somehow specific to this PR?

**Review by demiankatz — COMMENTED** [review]
One quick translation question. Also, some more generic responses to your questions:

1.) I'm not sure what the difference between "digital" and "electronic" is supposed to be. If that's an Alma-specific quirk, we might want to come up with a way to express it more generically, so we're not locked into concepts initiated by a particular platform.

2.) As you say, pagination is currently going to reload the entire holdings tab. It might make sense to create some kind of AJAX shortcut to get just the paged data without the other parts of the tab if we're going to have multiple independent lists. Are the digital/electronic holdings likely to need their own separate pagination, or will they always reliably fit on a single page?

(I haven't actually tried running this code yet -- currently I'm just trying to get some conversations started -- so I apologize if I'm misunderstanding what's going on here).

File: languages/fi.ini (hunk 1 above)
Comment (EreMaijala) [reply languages/fi.ini]: Specific to status display in Finnish, there were two columns with the same translation (this was being wrong)...

Comment (demiankatz) [reply languages/fi.ini]: Should I port this change over to the release-6.0 branch, then?

**Comment by EreMaijala** [comment]
@demiankatz 

1. I'm not quite sure either, but I suppose electronic material is more like e-journals etc. born-electronic while digital is like versions of physical material. But I think I'll remove the distinction from the code for now (apart from config in the Alma driver). It can be added later if necessary.

2.I agree. Then again, the page already does at least getRealTimeHistory(), so this just adds to it. I'd take the AJAX method to a separate issue. I don't think electronic/digital will need paging. At least the Alma availability API that returns the data doesn't have any.

**Comment by demiankatz** [comment]
@EreMaijala, sounds good, and I agree that AJAX work should be done separately from this. One problem at a time! I'll wait for your revisions (and perhaps feedback from @betullam) before delving deeper into this.

File: languages/fi.ini (hunk 1 above)
Comment (EreMaijala) [reply languages/fi.ini]: That'd make sense. Thanks!

**Comment by EreMaijala** [comment]
This should do it. Due to some setup issues I'm currently unable to verify with actual Alma connection.

File: languages/fi.ini (hunk 1 above)
Comment (demiankatz) [reply languages/fi.ini]: One last question before I apply the change: I notice that the language file still contains the line `By Call Number = "Sijainnin mukaan"` -- should this be adjusted to match?

Comment (demiankatz) [reply languages/fi.ini]: ...also `Multiple Call Numbers = "Useita sijainteja"`

Comment (demiankatz) [reply languages/fi.ini]: Maybe it's worth opening a separate PR to correct these three lines? Or you can just give me translations here and I'll fix it for you -- whatever is less work for you is fine with me!

**Comment by demiankatz** [comment]
Thanks, @EreMaijala. Do you expect that you will eventually be able to test this? In any case, I have just emailed @betullam to be sure he is receiving notifications about all of this work in progress; hopefully we'll get more feedback from him when time permits.

File: languages/fi.ini (hunk 1 above)
Comment (EreMaijala) [reply languages/fi.ini]: I think those are fine. "Sijainti" means "Location", and that's sort of shelf location.

**Comment by EreMaijala** [comment]
Yes, I certainly hope so! And I hope it happens soon..

File: languages/fi.ini (hunk 1 above)
Comment (demiankatz) [reply languages/fi.ini]: Okay, thanks for the confirmation. I've taken care of the merging of the fix!

File: languages/en.ini [hunk 2]
```diff
@@ -283,6 +283,7 @@ Description = "Description"
 Desired Username = "Desired Username"
 Detailed View = "Detailed View"
 Details = "Staff View"
+Digital = "Digital"
```
Comment (demiankatz) [inline languages/en.ini]: Now that the Digital distinction has been removed, can we also remove these translations?

Comment (EreMaijala) [reply languages/en.ini]: Right, of course. Now removed.

**Comment by betullam** [comment]
Unfortunately I think I will not be able to test this before my vacation which starts in 2 days and will last for 3 weeks. We don't (yet) use digital or electronic resources in Alma. All our digital stuff is added directly to VuFinds Solr index ... so it is not coming from Alma. For testing your changes I will have to set up some test cases in our Alma Sandbox which will take some time.

But for clarification of the terms "electronic" and "digital": I found this information in the Alma documentation:

> Digital resources are files that are owned by the institution, such as images, videos, sound recordings, and PDFs. Digital resources differ from electronic resources, which are resources to which the library subscribes.

Anyway, I also think that its not necessary yet to distinguish between electronic and digital. As Ere mentioned: it can be added later if necessary.

**Comment by demiankatz** [comment]
Thanks, @betullam. If you don't object, I may merge this without your testing (once @EreMaijala is comfortable with it), since it is a new feature that has no risk of breaking existing functionality, and we can bug-fix it later as needed. I also agree that based on this definition, the end user probably does not care about the digital/electronic distinction -- it sounds more like a matter of internal significance for resource management. Thanks!

**Comment by betullam** [comment]
@demiankatz Yes that's OK for me, no problem.

**Comment by EreMaijala** [comment]
@demiankatz I was able to test now. I'm good to go.

File: themes/bootstrap3/templates/RecordTab/holdingsils/electronic.phtml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+<h3><?=$this->transEsc('Electronic')?></h3>
+<table class="table table-striped">
+  <?php $i = 0; foreach ($this->holdings as $item): ?>
```
Comment (demiankatz) [inline themes/bootstrap3/templates/RecordTab/holdingsils/electronic.phtml]: It appears that `$i` is unused. Can we remove it?

File: themes/bootstrap3/templates/RecordTab/holdingsils.phtml [hunk 4]
```diff
@@ -12,7 +12,12 @@
     try {
         $holdings = $this->driver->getRealTimeHoldings();
     } catch (\VuFind\Exception\ILS $e) {
-        $holdings = ['holdings' => []];
+        $holdings = [
+          'holdings' => [],
+          'total' => 0,
```
Comment (demiankatz) [inline themes/bootstrap3/templates/RecordTab/holdingsils.phtml]: Should we initialize electronic_holdings here too?

File: module/VuFind/src/VuFind/ILS/Driver/Alma.php [hunk 5]
```diff
@@ -1549,6 +1515,125 @@ public function parseDate($date, $withTime = false)
...
+                    // Digital
+                    $deliveryUrl
+                        = $this->config['Holdings']['digitalDeliveryUrl'] ?? '';
+                    $digitalItems = $record->getFields('AVD');
+                    foreach ($digitalItems as $field) {
+                        $item = $tmpl;
+                        unset($item['callnumber']);
+                        $item['availability'] = true;
+                        $item['location'] = $field->getSubfield('e')->getData();
+                        if ($deliveryUrl) {
```
Comment (demiankatz) [inline module/VuFind/src/VuFind/ILS/Driver/Alma.php]: Should we issue a warning in some fashion if the URL is unset? Seems like an easily missed configuration detail.

**Review by demiankatz — COMMENTED** [review]
@EreMaijala, I gave this a closer look before merging and noticed a few minor things... Also, we should update the wiki when merging to document the addition to the ILS driver spec. Are the contents of electronic_holdings identical to the contents of holdings? At a glance, it appears so, but I didn't look too closely to check for electronic-specific details. In any case, I'll add a TODO checkbox about updating the wiki, and I'll take care of it when I get clarification and the go-ahead from you. Thanks!

File: themes/bootstrap3/templates/RecordTab/holdingsils/electronic.phtml (hunk 3 above)
Comment (EreMaijala) [reply themes/bootstrap3/templates/RecordTab/holdingsils/electronic.phtml]: Oops. Yes.

File: themes/bootstrap3/templates/RecordTab/holdingsils.phtml (hunk 4 above)
Comment (EreMaijala) [reply themes/bootstrap3/templates/RecordTab/holdingsils.phtml]: It's checked with !empty() but it doesn't hurt to initialize it here too.

File: module/VuFind/src/VuFind/ILS/Driver/Alma.php (hunk 5 above)
Comment (EreMaijala) [reply module/VuFind/src/VuFind/ILS/Driver/Alma.php]: Good idea, I'll add one.

File: module/VuFind/src/VuFind/ILS/Driver/Alma.php [hunk 6]
```diff
@@ -1549,6 +1515,132 @@ public function parseDate($date, $withTime = false)
...
+                            $item['locationhref'] = str_replace(
+                                '%%id%%',
+                                $field->getSubfield('b')->getData(),
+                                $deliveryUrl
+                            );
+                        }
+                        $item['type'] = 'digital';
```
[the remark this answers is not in the document]
Comment (EreMaijala) [reply module/VuFind/src/VuFind/ILS/Driver/Alma.php]: Right, that was a leftover. Now removed.

**Review by demiankatz — APPROVED** [review]

**Comment by EreMaijala** [comment]
@demiankatz I blew it for the Demo driver. Will fix..

**Comment by demiankatz** [comment]
Thanks again -- all makes sense now. Merging!

**Pull request merged**